build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -1288,6 +1670,228 @@ message WaitExecutionRequest {
+  // [Command][build.bazel.remote.execution.v2.Command] but the
+  // `input_root_digest` can refer to a subset of the files that are predicted
+  // to be needed to run the `Action`. The input root can contain all files
+  // needed for the command, in which case the action fingerprint should lead
+  // to just one output (or none if there are no cache hits). When the input
+  // root contains fewer files than those actually needed by the action, the
+  // action fingerprint gets less specific, and the number of possible matches
+  // grows. It is possible for the input root to contain zero files, which is
+  // a very non-specific fingerprint as it can match any action results that
+  // match the `Command`. Less-specific action fingerprints lead to more I/O
+  // and time spent matching `ObservedPathSet`s against the client filesystem;
+  // adding in any predictions about the filesystem "focuses" the lookup to
+  // more-specific results.
+  Digest action_fingerprint_digest = 2;
+
+  // A hint to the server to request inlining `ObservedPathSet`s in the
+  // [GetPotentialActionResultsResponse][build.bazel.remote.execution.v2.GetPotentialActionResultsResponse]
+  // message.
+  bool inline_pathsets = 3;
+
+  // A maximum page size to request. If present, the server will respond with no more
+  // than this many items. Regardless of whether a page size is specified, the
+  // server may place its own limit on the number of items to be returned and
+  // require the client to retrieve more items using a subsequent request.
+  int32 page_size = 4;
+
+  // A page token, which must be a value received in a previous
+  // [GetPotentialActionResultsResponse][build.bazel.remote.execution.v2.GetPotentialActionResultsResponse].
+  // If present, the server will use that token as an offset, returning only
+  // that page and the ones that succeed it.
+  string page_token = 5;
+}
+
+// Returns a set of zero or more
+// (pathset, observed-input digest, `ActionResult` digest) results
+// from a two-phase cache lookup call to
+// [ActionCache.GetPotentialActionResults][build.bazel.remote.execution.v2.ActionCache.GetPotentialActionResults].
+// These results are compared to the local filesystem and, on a match to the
+// observed-input digest, the associated `ActionResult` is used to retrieve
+// cached outputs.
+message GetPotentialActionResultsResponse {
+  // A set of possible matches to the client filesystem along with a reference to build outputs.
+  repeated ObservedPathSetAndInputDigest potential_action_matches = 1;
+
+  // If present, signifies that there are more results which the client can
+  // retrieve by passing this as the page_token in a subsequent
+  // [request][build.bazel.remote.execution.v2.GetPotentialActionResults].
+  // If empty, signifies that this is the last page of results.
+  string next_page_token = 2;
+}
+
+// The ordered file reads and probes and directory enumerations used in
```

Comment:
More discussion of what is actually being encoded is needed. This ordered list of operations is representing what, exactly?